src/graphics/vertex-format.js

```diff
@@ -75,28 +79,41 @@ Object.assign(pc, function () {
         this.hasColor = false;
         this.hasTangents = false;
         this._defaultInstancingFormat = null;
+        this.verticesByteSize = 0;
+        this.vertexCount = vertexCount;
+        this.interleaved = !vertexCount;
 
-        // calculate total size
+        // calculate total size of the vertex
         this.size = description.reduce(function (total, desc) {
             return total + Math.ceil(desc.components * _typeSize[desc.type] / 4) * 4;
         }, 0);
 
-        var offset = 0;
+        var offset = 0, elementSize;
         for (i = 0, len = description.length; i < len; i++) {
             var elementDesc = description[i];
+
+            // align up the offset to elementSize (when vertexCount is specified only - case of non-interleaved format)
+            elementSize = elementDesc.components * _typeSize[elementDesc.type];
```

Comment:
This calculation needs to use the same logic as this.size calculation above. All elements of a vertex must be aligned to 4byte boundaries, whether interleaved or not.

Reply:
I think you are right .. this is existing bug for interleaved format, but it's a good time to fix it perhaps.

Reply:
looking at the code .. I think in case user manually creates interleaved format, he's expected to correctly provide 'offset' and 'stride' parameters in case an alignment is needed. I don't think we should apply any alignment on top of provided values?

Reply:
I had thought vertex attributes themselves should ideally be 4 byte aligned (not just the vertex attribute offset), though I may be wrong. For example I assumed an array of RGB888 should not be stored tightly packed as RGBRGBRGB but rather as RGB_RGB_RGB_ where _ is a padding byte.

See https://developer.apple.com/library/archive/documentation/3DDrawing/Conceptual/OpenGLES_ProgrammingGuide/TechniquesforWorkingwithVertexData/TechniquesforWorkingwithVertexData.html#//apple_ref/doc/uid/TP40008793-CH107-SW1 section 'Avoid Misaligned Vertex Data'

The interleaved data does this correctly because the offset calc takes into account 4-byte boundary.

Reply:
Webgl supports it, Apple does not like it for performance.
Considering this is probably not likely to happen, perhaps I should log a warning in debug mode when format is created to warn of possible performance impact, but use it anyways. It feels wrong to have to interleave data with gaps for non-interleaved format.

Reply:
Yeah it could be updated later if necessary.